File: solutions/typescript/triangle/1/triangle.ts
export class Triangle {
  private sides: number[];

  constructor(...sides: number[]) {
    this.sides = sides;
  }

  private isValidTriangle(): boolean {
    const [a, b, c] = this.sides;
    
    // All sides must have positive length
    if (a <= 0 || b <= 0 || c <= 0) {
      return false;
    }
    
    // Triangle inequality theorem
    return a + b >= c && b + c >= a && a + c >= b;
  }

  get isEquilateral(): boolean {
    if (!this.isValidTriangle()) {
      return false;
    }
    
    const [a, b, c] = this.sides;
    return a === b && b === c;
  }

  get isIsosceles(): boolean {
    if (!this.isValidTriangle()) {
      return false;
    }
    
    const [a, b, c] = this.sides;
    return a === b || b === c || a === c;
  }

  get isScalene(): boolean {
    if (!this.isValidTriangle()) {
      return false;
    }
    
    const [a, b, c] = this.sides;
    return a !== b && b !== c && a !== c;
  }
}

File: solutions/java/triangle/1/src/main/java/Triangle.java
class Triangle {
    private final double side1;
    private final double side2;
    private final double side3;
    
    Triangle(double side1, double side2, double side3) throws TriangleException {
        // Validate all sides are positive
        if (side1 <= 0 || side2 <= 0 || side3 <= 0) {
            throw new TriangleException();
        }
        
        // Validate triangle inequality: sum of any two sides must be >= third side
        if (side1 + side2 < side3 || side2 + side3 < side1 || side1 + side3 < side2) {
            throw new TriangleException();
        }
        
        this.side1 = side1;
        this.side2 = side2;
        this.side3 = side3;
    }

    boolean isEquilateral() {
        return side1 == side2 && side2 == side3;
    }

    boolean isIsosceles() {
        // At least two sides equal (includes equilateral triangles)
        return side1 == side2 || side2 == side3 || side1 == side3;
    }

    boolean isScalene() {
        // All sides different
        return side1 != side2 && side2 != side3 && side1 != side3;
    }
}
